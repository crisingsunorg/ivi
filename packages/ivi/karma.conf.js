const webpack = require("webpack");

module.exports = function (config) {
  config.set({
    browsers: ["ChromeHeadless"],
    frameworks: ["mocha"],
    reporters: ["mocha"],
    preprocessors: { "tests/index.ts": ["webpack", "sourcemap"] },
    files: ["tests/index.ts"],

    colors: true,
    autoWatch: true,

    webpack: {
      module: {
        rules: [
          {
            test: /\.ts$/,
            exclude: /node_modules/,
            use: [
              {
                loader: "ts-loader",
                options: {
                  configFileName: "tsconfig.tests.json",
                },
              },
            ],
          },
        ],
      },
      plugins: [
        new webpack.DefinePlugin({
          "__IVI_DEV__": true,
          "__IVI_BROWSER__": true,
        }),
        new webpack.SourceMapDevToolPlugin({
          test: /\.(ts|js)$/,
        }),
      ],
      resolve: {
        extensions: [".ts", ".js"],
      },
      performance: {
        hints: false
      },
    },

    webpackMiddleware: {
      stats: "errors-only",
      noInfo: true
    },

    mochaReporter: {
      showDiff: true,
    },

    client: {
      mocha: {
        reporter: "html",
        ui: "bdd",
      }
    },

    mime: {
      "text/x-typescript": ["ts"],
    },
  });
};
