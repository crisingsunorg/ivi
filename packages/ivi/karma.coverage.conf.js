const path = require("path");
const webpack = require("webpack");
const merge = require("webpack-merge");
const baseConfig = require("./karma.conf");

module.exports = function (config) {
  baseConfig(config);

  config.set({
    singleRun: true,

    webpack: merge.smart(config.webpack, {
      module: {
        rules: [{
          test: /\.ts$/,
          include: path.resolve("src/"),
          exclude: [
            /node_modules/,
            path.resolve("src/dev_mode/dev_mode.ts"),
            path.resolve("src/dev_mode/stack_trace.ts"),
          ],
          loader: "istanbul-instrumenter-loader",
          enforce: "post",
        }],
      },
    }),

    reporters: ["coverage-istanbul"],

    coverageIstanbulReporter: {
      reports: ["lcov", "text"],
      dir: "./coverage",
      fixWebpackSourcePath: true,
      skipFilesWithNoCoverage: true,
    },
  });

  if (process.env.TRAVIS) {
    config.coverageIstanbulReporter.reports = ["lcov"];
  }
};
