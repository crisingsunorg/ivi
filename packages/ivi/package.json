{
  "name": "ivi",
  "main": "dist/main/index.js",
  "module": "dist/module/index.js",
  "es2016": "dist/es2016/index.js",
  "unpkg": "dist/unpkg/ivi.js",
  "typings": "dist/typings/index.d.ts",
  "sideEffects": false,
  "version": "0.10.0",
  "description": "Javascript (TypeScript) library for building web user interfaces.",
  "license": "MIT",
  "author": {
    "name": "Boris Kaul",
    "email": "localvoid@gmail.com",
    "url": "https://github.com/localvoid"
  },
  "keywords": [
    "ui",
    "virtualdom",
    "vdom",
    "diff",
    "browser",
    "typescript"
  ],
  "homepage": "https://github.com/ivijs/ivi",
  "repository": {
    "type": "git",
    "url": "https://github.com/ivijs/ivi"
  },
  "bugs": {
    "url": "https://github.com/ivijs/ivi/issues"
  },
  "github": "https://github.com/ivijs/ivi",
  "files": [
    "dist",
    "src",
    "README.md",
    "LICENSE"
  ],
  "scripts": {
    "prepublishOnly": "npm run clean && npm run dist",
    "clean": "rm -rf ./dist",
    "dist": "npm run dist:main && npm run dist:module && npm run dist:es2016 && npm run dist:unpkg",
    "dist:main": "../../node_modules/.bin/tsc -p ./tsconfig.build.main.json",
    "dist:module": "../../node_modules/.bin/tsc -p ./tsconfig.build.module.json",
    "dist:es2016": "../../node_modules/.bin/tsc -p ./tsconfig.build.es2016.json",
    "dist:unpkg": "../../node_modules/.bin/webpack --config ./webpack.config.unpkg.js"
  },
  "dependencies": {},
  "peerDependencies": {
    "ivi-core": "^0.10.0",
    "ivi-scheduler": "^0.10.0",
    "ivi-events": "^0.10.0"
  },
  "devDependencies": {
    "ivi-core": "^0.10.0",
    "ivi-scheduler": "^0.10.0",
    "ivi-events": "^0.10.0"
  }
}
