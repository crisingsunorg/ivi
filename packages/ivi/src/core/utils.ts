/**
 * mut mutates an object.
 *
 * @example
 *
 *     interface Entry {
 *       _dirty: number;
 *       title: string;
 *     }
 *     const m = (entry: Entry) => { entry._dirty = dirty(); };
 *     const useEntry = selector((entry: Entry) => entry._dirty);
 *     const entrySetTitle = mut(m, (entry: Entry, title: string) => { entry.title = title; });
 *
 *     const EntryView = component<Entry>((c) => {
 *       const dirtyCheckEntry = useEntry(c);
 *
 *       return (entry) => (
 *         dirtyCheckEntry(entry),
 *
 *         div().t(entry.title)
 *       )
 *     });
 *
 * @param s - Selector function.
 * @param shouldUpdate - Should update function.
 * @returns selector factory.
 */
export const mut = <T, U extends any[], V>(d: (v: T, flags?: V) => void, fn: (v: T, ...args: U) => void | V) => (
  function () {
    const r = fn.apply(void 0, arguments);
    if (r === void 0 || r > 0) {
      d(arguments[0], r);
    }
  } as (v: T, ...args: U) => void | V
);
