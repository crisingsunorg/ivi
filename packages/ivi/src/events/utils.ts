import { PASSIVE_EVENTS } from "ivi-core";

/* istanbul ignore next */
/**
 * getEventTarget retrieves target element from an event.
 *
 * There are still some differences in modern browsers, so we need to use this function to retrieve event targets.
 *
 * #quirks
 *
 * @param ev Event.
 * @returns Target Element.
 */
export function getEventTarget(ev: Event): EventTarget {
  let target = ev.target || window;

  /**
   * Some browsers are implementing it according to SVG 1.1 specs:
   *
   * https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/7998724/
   *
   * "If event attributes are assigned to referenced elements, then the actual target for the event will be the
   * SVGElementInstance object within the 'instance tree' corresponding to the given referenced element."
   *
   * https://www.w3.org/TR/SVG/struct.html#UseElement
   *
   * SVG 2 redefined the use event handling model:
   *
   * - Removed the SVGElementInstance and SVGElementInstanceList interfaces, and the corresponding attributes on the
   *   SVGUseElement interface.
   * - Changed the ‘use’ element event flow to follow the Shadow DOM spec.
   *
   * https://www.w3.org/TR/SVG2/changes.html#structure
   */
  if ((target as any).correspondingUseElement !== undefined) {
    target = (target as any).correspondingUseElement;
  }

  return target;
}

/* istanbul ignore next */
/**
 * `{ capture: true, passive: true }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_CAPTURE_PASSIVE_OPTIONS = PASSIVE_EVENTS ? { "capture": true, "passive": true } : true;

/* istanbul ignore next */
/**
 * `{ capture: true, passive: false }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_CAPTURE_ACTIVE_OPTIONS = PASSIVE_EVENTS ? { "capture": true, "passive": false } : true;

/* istanbul ignore next */
/**
 * `{ passive: true }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_PASSIVE_OPTIONS = PASSIVE_EVENTS ? { "passive": true } : false;

/* istanbul ignore next */
/**
 * `{ passive: false }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_ACTIVE_OPTIONS = PASSIVE_EVENTS ? { "passive": false } : false;
