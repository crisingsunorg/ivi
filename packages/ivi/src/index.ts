/**
 * Common types and functions re-exported from ivi-core.
 */
export { KeyCode, KeyLocation, MouseButtons, SyncableValue, PROPERTY, UNSAFE_HTML, EVENT, _ } from "ivi-core";

// Virtual DOM
export { VNodeFlags } from "./vdom/flags";
export { StatelessComponent } from "./vdom/stateless_component";
export { StatefulComponent, Component, isComponentAttached } from "./vdom/stateful_component";
export { ConnectDescriptor } from "./vdom/connect_descriptor";
export {
  VNode,
  getDOMNode, getComponent,
  stopDirtyChecking,
} from "./vdom/vnode";
export { fragment, map, mapRange, mapIterable } from "./vdom/vnode_collections";
export { element } from "./vdom/element";
export { statefulComponent, statelessComponent, withShouldUpdate, context, connect } from "./vdom/vnode_factories";
export { invalidate, dirtyCheck, render } from "./vdom/scheduler";

// Events
export {
  EventHandlerFlags, SyntheticEventFlags, EventFlags, PREVENT_DEFAULT, STOP_PROPAGATION, NativeEventSourceFlags,
} from "./events/flags";
export {
  getEventTarget,
  EVENT_CAPTURE_PASSIVE_OPTIONS, EVENT_CAPTURE_ACTIVE_OPTIONS, EVENT_PASSIVE_OPTIONS, EVENT_ACTIVE_OPTIONS,
} from "./events/utils";
export { SyntheticEvent } from "./events/synthetic_event";
export { EventDispatcher } from "./events/event_dispatcher";
export { EventHandler } from "./events/event_handler";
export { DispatchTarget } from "./events/dispatch_target";
export { accumulateDispatchTargets } from "./events/accumulate_dispatch_targets";
export { dispatchEvent } from "./events/dispatch_event";
export { SyntheticNativeEvent } from "./events/synthetic_native_event";
export {
  NativeEventDispatcher, createNativeEventDispatcher,
  beforeNativeEvent, afterNativeEvent,
  removeBeforeNativeEvent, removeAfterNativeEvent,
} from "./events/native_event_dispatcher";
export {
  EVENT_DISPATCHER_ABORT, EVENT_DISPATCHER_ACTIVATE, EVENT_DISPATCHER_ARIA_REQUEST, EVENT_DISPATCHER_BEFORE_ACTIVATE,
  EVENT_DISPATCHER_BEFORE_COPY, EVENT_DISPATCHER_BEFORE_CUT, EVENT_DISPATCHER_BEFORE_DEACTIVATE,
  EVENT_DISPATCHER_BEFORE_PASTE, EVENT_DISPATCHER_BLUR, EVENT_DISPATCHER_CAN_PLAY, EVENT_DISPATCHER_CAN_PLAYTHROUGH,
  EVENT_DISPATCHER_CHANGE, EVENT_DISPATCHER_CLICK, EVENT_DISPATCHER_CONTEXT_MENU, EVENT_DISPATCHER_COPY,
  EVENT_DISPATCHER_CUE_CHANGE, EVENT_DISPATCHER_CUT, EVENT_DISPATCHER_DOUBLE_CLICK, EVENT_DISPATCHER_DEACTIVATE,
  EVENT_DISPATCHER_DRAG, EVENT_DISPATCHER_DRAG_END, EVENT_DISPATCHER_DRAG_ENTER, EVENT_DISPATCHER_DRAG_LEAVE,
  EVENT_DISPATCHER_DRAG_OVER, EVENT_DISPATCHER_DRAG_START, EVENT_DISPATCHER_DROP, EVENT_DISPATCHER_DURATION_CHANGE,
  EVENT_DISPATCHER_EMPTIED, EVENT_DISPATCHER_ENCRYPTED, EVENT_DISPATCHER_ENDED, EVENT_DISPATCHER_ERROR,
  EVENT_DISPATCHER_FOCUS, EVENT_DISPATCHER_GOT_POINTER_CAPTURE, EVENT_DISPATCHER_INPUT, EVENT_DISPATCHER_INVALID,
  EVENT_DISPATCHER_KEY_DOWN, EVENT_DISPATCHER_KEY_PRESS, EVENT_DISPATCHER_KEY_UP, EVENT_DISPATCHER_LOAD,
  EVENT_DISPATCHER_LOADED_DATA, EVENT_DISPATCHER_LOADED_METADATA, EVENT_DISPATCHER_LOAD_START,
  EVENT_DISPATCHER_LOST_POINTER_CAPTURE, EVENT_DISPATCHER_MOUSE_DOWN, EVENT_DISPATCHER_MOUSE_ENTER,
  EVENT_DISPATCHER_MOUSE_LEAVE, EVENT_DISPATCHER_MOUSE_MOVE, EVENT_DISPATCHER_MOUSE_OUT, EVENT_DISPATCHER_MOUSE_OVER,
  EVENT_DISPATCHER_MOUSE_UP, EVENT_DISPATCHER_PASTE, EVENT_DISPATCHER_PAUSE, EVENT_DISPATCHER_PLAY,
  EVENT_DISPATCHER_PLAYING, EVENT_DISPATCHER_POINTER_CANCEL, EVENT_DISPATCHER_POINTER_DOWN,
  EVENT_DISPATCHER_POINTER_ENTER, EVENT_DISPATCHER_POINTER_LEAVE, EVENT_DISPATCHER_POINTER_MOVE,
  EVENT_DISPATCHER_POINTER_OUT, EVENT_DISPATCHER_POINTER_OVER, EVENT_DISPATCHER_POINTER_UP, EVENT_DISPATCHER_PROGRESS,
  EVENT_DISPATCHER_RATE_CHANGE, EVENT_DISPATCHER_RESET, EVENT_DISPATCHER_SCROLL, EVENT_DISPATCHER_SEEKED,
  EVENT_DISPATCHER_SEEKING, EVENT_DISPATCHER_SELECT, EVENT_DISPATCHER_SELECT_START, EVENT_DISPATCHER_STALLED,
  EVENT_DISPATCHER_SUBMIT, EVENT_DISPATCHER_SUSPEND, EVENT_DISPATCHER_TIME_UPDATE, EVENT_DISPATCHER_TOUCH_CANCEL,
  EVENT_DISPATCHER_TOUCH_END, EVENT_DISPATCHER_TOUCH_MOVE, EVENT_DISPATCHER_TOUCH_START, EVENT_DISPATCHER_UNLOAD,
  EVENT_DISPATCHER_VOLUME_CHANGE, EVENT_DISPATCHER_WAITING, EVENT_DISPATCHER_WHEEL, EVENT_DISPATCHER_ACTIVE_TOUCH_END,
  EVENT_DISPATCHER_ACTIVE_TOUCH_MOVE, EVENT_DISPATCHER_ACTIVE_TOUCH_START, EVENT_DISPATCHER_ACTIVE_WHEEL,

  createNativeEventHandler,

  onAbort, onActivate, onAriaRequest, onBeforeActivate, onBeforeCopy, onBeforeCut, onBeforeDeactivate, onBeforePaste,
  onBlur, onCanPlay, onCanPlaythrough, onChange, onClick, onContextMenu, onCopy, onCueChange, onCut, onDoubleClick,
  onDeactivate, onDrag, onDragEnd, onDragEnter, onDragLeave, onDragOver, onDragStart, onDrop, onDurationChange,
  onEmptied, onEncrypted, onEnded, onError, onFocus, onGotPointerCapture, onInput, onInvalid, onKeyDown, onKeyPress,
  onKeyUp, onLoad, onLoadedData, onLoadedMetadata, onLoadStart, onLostPointerCapture, onMouseDown, onMouseEnter,
  onMouseLeave, onMouseMove, onMouseOut, onMouseOver, onMouseUp, onPaste, onPause, onPlay, onPlaying, onPointerCancel,
  onPointerDown, onPointerEnter, onPointerLeave, onPointerMove, onPointerOut, onPointerOver, onPointerUp, onProgress,
  onRateChange, onReset, onScroll, onSeeked, onSeeking, onSelect, onSelectStart, onStalled, onSubmit, onSuspend,
  onTimeUpdate, onTouchCancel, onTouchEnd, onTouchMove, onTouchStart, onUnload, onVolumeChange, onWaiting, onWheel,
  onActiveTouchEnd, onActiveTouchMove, onActiveTouchStart, onActiveWheel,
} from "./events/native_events";
