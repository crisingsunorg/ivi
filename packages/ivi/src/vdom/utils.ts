import { Box } from "ivi-shared";
import { OpState } from "./state";
import { Component } from "./component";
import { useSelect } from "./hooks";
import { getDOMNode } from "./reconciler";

/**
 * selector creates a selector factory.
 *
 * @example
 *
 *     interface Entry {
 *       title: string;
 *     }
 *     const useEntryTitle = selector((entry: Entry) => entry.title);
 *
 *     const EntryView = component<Entry>((c) => {
 *       const getEntryTitle = useEntryTitle(c);
 *
 *       return (entry) => div(_, _, getEntryTitle(entry));
 *     });
 *
 * @typename T Selector value type.
 * @param s Selector function.
 * @returns Selector factory.
 */
export function selector<T>(
  s: () => T,
): (component: Component) => () => T;

/**
 * selector creates a selector factory.
 *
 * @example
 *
 * interface Entry {
 *   title: string;
 * }
 *     const useEntryTitle = selector((entry: Entry) => entry.title);
 *
 *     const EntryView = component<Entry>((c) => {
 *       const getEntryTitle = useEntryTitle(c);
 *
 *       return (entry) => div(_, _, getEntryTitle(entry));
 *     });
 *
 * @typename T Selector value type.
 * @param s Selector function.
 * @param shouldUpdate Should update function.
 * @returns Selector factory.
 */
export function selector<T, P>(
  s: (props: P, prev?: T | undefined) => T,
  shouldUpdate?: undefined extends P ? undefined : (prev: P, next: P) => boolean,
): (component: Component) => undefined extends P ? () => T : (props: P) => T;

export function selector<T, P>(
  s: (props: P, prev?: T | undefined) => T,
  shouldUpdate?: undefined extends P ? undefined : (prev: P, next: P) => boolean,
): (component: Component) => undefined extends P ? () => T : (props: P) => T {
  return (component: Component) => useSelect(component, s, shouldUpdate);
}

/**
 * findDOMNode finds a closest DOM node.
 *
 * @typeparam T DOM Node type.
 * @param box Boxed op node state.
 * @returns DOM Node or a `null` value.
 */
export const findDOMNode = <T extends Node>(
  box: Box<OpState | null>,
) => box.v === null ? null : getDOMNode(box.v) as T;
