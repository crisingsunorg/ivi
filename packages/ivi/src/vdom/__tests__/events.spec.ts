import { EventFlags, EventHandler, SyntheticEvent } from "ivi";
import * as Events from "ivi";
import * as h from "ivi-html";
import { startRender } from "./utils";

function createMouseEvent(type: string): MouseEvent {
  if (document.createEvent) {
    const ev = document.createEvent("MouseEvent");
    ev.initMouseEvent(type, true, true, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null!);
    return ev;
  }
  return new MouseEvent(type);
}

export interface EventCounter {
  value: number;
  event: EventHandler<any>;
}

export function eventCounter(handlerFactory: (
  handler: (ev: SyntheticEvent) => void,
  capture?: boolean) => EventHandler<any>,
): EventCounter {
  const c = {
    value: 0,
    event: null as EventHandler<any> | null,
  };
  c.event = handlerFactory(() => {
    c.value++;
  }, false);

  return c as EventCounter;
}

describe("events", () => {
  const container = document.createElement("div");
  document.body.appendChild(container);

  test(`<div onclick=FN>`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      const n = r(h.div().e([click.event]));
      n.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(1);
    });
  });

  test(`<div onclick=FN onclick=FN>`, () => {
    startRender((r) => {
      const aClick = eventCounter(Events.onClick);
      const bClick = eventCounter(Events.onClick);
      const n = r(h.div().e([aClick.event, bClick.event]));
      n.dispatchEvent(createMouseEvent("click"));

      expect(aClick.value).toBe(1);
      expect(bClick.value).toBe(1);
    });
  });

  test(`<div onclick=FN onmousedown=FN>`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      const mousedown = eventCounter(Events.onMouseDown);
      const n = r(h.div().e([click.event, mousedown.event]));

      n.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(1);
      expect(mousedown.value).toBe(0);

      n.dispatchEvent(createMouseEvent("mousedown"));

      expect(click.value).toBe(1);
      expect(mousedown.value).toBe(1);
    });
  });

  test(`unassigned => []`, () => {
    startRender((r) => {
      expect(() => {
        r(h.div());
        r(h.div().e([]));
      }).not.toThrow();
    });
  });

  test(`[] => unassigned`, () => {
    startRender((r) => {
      expect(() => {
        r(h.div().e([]));
        r(h.div());
      }).not.toThrow();
    });
  });

  test(`null => []`, () => {
    startRender((r) => {
      expect(() => {
        r(h.div().e(null));
        r(h.div().e([]));
      }).not.toThrow();
    });
  });

  test(`[] => null`, () => {
    startRender((r) => {
      expect(() => {
        r(h.div().e([]));
        r(h.div().e(null));
      }).not.toThrow();
    });
  });

  test(`[] => []`, () => {
    startRender((r) => {
      expect(() => {
        r(h.div().e([]));
        r(h.div().e([]));
      }).not.toThrow();
    });
  });

  test(`unassigned => onclick`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div());
      const b = r(h.div().e(
        click.event,
      ));
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(1);
    });
  });

  test(`onclick => unassigned`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div().e(
        click.event,
      ));
      const b = r(h.div());
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(0);
    });
  });

  test(`null => onclick`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div().e(
        null,
      ));
      const b = r(h.div().e(
        click.event,
      ));
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(1);
    });
  });

  test(`onclick => null`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div().e(
        click.event,
      ));
      const b = r(h.div().e(
        null,
      ));
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(0);
    });
  });

  test(`unassigned => [onclick]`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div());
      const b = r(h.div().e([
        click.event,
      ]));
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(1);
    });
  });

  test(`[onclick] => unassigned`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div().e([
        click.event,
      ]));
      const b = r(h.div());
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(0);
    });
  });

  test(`null => [onclick]`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div().e(null));
      const b = r(h.div().e([
        click.event,
      ]));
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(1);
    });
  });

  test(`[onclick] => null`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div().e([
        click.event,
      ]));
      const b = r(h.div().e(
        null,
      ));
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(0);
    });
  });

  test(`onclick => [onclick]`, () => {
    startRender((r) => {
      const click1 = eventCounter(Events.onClick);
      const click2 = eventCounter(Events.onClick);
      r(h.div().e(
        click1.event,
      ));
      const b = r(h.div().e([
        click2.event,
      ]));
      b.dispatchEvent(createMouseEvent("click"));

      expect(click1.value).toBe(0);
      expect(click2.value).toBe(1);
    });
  });

  test(`[onclick] => onclick`, () => {
    startRender((r) => {
      const click1 = eventCounter(Events.onClick);
      const click2 = eventCounter(Events.onClick);
      r(h.div().e([
        click1.event,
      ]));
      const b = r(h.div().e(
        click2.event,
      ));
      b.dispatchEvent(createMouseEvent("click"));

      expect(click1.value).toBe(0);
      expect(click2.value).toBe(1);
    });
  });

  test(`[] => [onclick]`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div().e([]));
      const b = r(h.div().e([
        click.event,
      ]));
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(1);
    });
  });

  test(`[onclick] => []`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div().e([
        click.event,
      ]));
      const b = r(h.div().e([]));
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(0);
    });
  });

  test(`unassigned => [onclick, onmousedown]`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      const mousedown = eventCounter(Events.onMouseDown);
      r(h.div());
      const b = r(h.div().e([click.event, mousedown.event]));

      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(1);

      b.dispatchEvent(createMouseEvent("mousedown"));

      expect(mousedown.value).toBe(1);
    });
  });

  test(`null => [onclick, onmousedown]`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      const mousedown = eventCounter(Events.onMouseDown);
      r(h.div().e(null));
      const b = r(h.div().e([
        click.event,
        mousedown.event,
      ]));

      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(1);

      b.dispatchEvent(createMouseEvent("mousedown"));

      expect(mousedown.value).toBe(1);
    });
  });

  test(`[] => [onclick, onmousedown]`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      const mousedown = eventCounter(Events.onMouseDown);
      r(h.div().e([]));
      const b = r(h.div().e([
        click.event,
        mousedown.event,
      ]));

      b!.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(1);

      b.dispatchEvent(createMouseEvent("mousedown"));

      expect(mousedown.value).toBe(1);
    });
  });

  test(`null => [onclick, onclick]`, () => {
    startRender((r) => {
      const aClick = eventCounter(Events.onClick);
      const bClick = eventCounter(Events.onClick);
      r(h.div().e(null));
      const b = r(h.div().e([
        aClick.event,
        bClick.event,
      ]));
      b.dispatchEvent(createMouseEvent("click"));

      expect(aClick.value).toBe(1);
      expect(bClick.value).toBe(1);
    });
  });

  test(`[onclick] => [onclick]`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div().e([
        click.event,
      ]));
      const b = r(h.div().e([click.event]));
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(1);
    });
  });

  test(`[onclick] => unassigned`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div().e([
        click.event,
      ]));
      const b = r(h.div());
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(0);
    });
  });

  test(`[onclick] => null`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div().e([
        click.event,
      ]));
      const b = r(h.div().e(
        null,
      ));
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(0);
    });
  });

  test(`[onclick] => []`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      r(h.div().e([
        click.event,
      ]));
      const b = r(h.div().e([]));
      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(0);
    });
  });

  test(`[onclick, null] => [null, onclick]`, () => {
    startRender((r) => {
      const aClick = eventCounter(Events.onClick);
      const bClick = eventCounter(Events.onClick);
      r(h.div().e([
        aClick.event,
        null,
      ]));
      const b = r(h.div().e([
        null,
        bClick.event,
      ]));
      b.dispatchEvent(createMouseEvent("click"));

      expect(aClick.value).toBe(0);
      expect(bClick.value).toBe(1);
    });
  });

  test(`[onclick, onmousedown] => []`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      const mousedown = eventCounter(Events.onMouseDown);
      r(h.div().e([
        click.event,
        mousedown.event,
      ]));
      const b = r(h.div().e([]));

      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(0);

      b.dispatchEvent(createMouseEvent("mousedown"));

      expect(mousedown.value).toBe(0);
    });
  });

  test(`[onclick, onmousedown] => null`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      const mousedown = eventCounter(Events.onMouseDown);
      r(h.div().e([
        click.event,
        mousedown.event,
      ]));
      const b = r(h.div().e(
        null,
      ));

      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(0);

      b.dispatchEvent(createMouseEvent("mousedown"));

      expect(mousedown.value).toBe(0);
    });
  });

  test(`[onclick, onmousedown] => [onclick]`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      const mousedown = eventCounter(Events.onMouseDown);
      r(h.div().e([
        click.event,
        mousedown.event,
      ]));
      const b = r(h.div().e([
        click.event,
      ]));

      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(1);

      b.dispatchEvent(createMouseEvent("mousedown"));

      expect(mousedown.value).toBe(0);
    });
  });

  test(`[onclick, onmousedown] => [onclick, onmouseup]`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      const mousedown = eventCounter(Events.onMouseDown);
      const mouseup = eventCounter(Events.onMouseUp);
      r(h.div().e([
        click.event,
        mousedown.event,
      ]));
      const b = r(h.div().e([
        click.event,
        mouseup.event,
      ]));

      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(1);

      b.dispatchEvent(createMouseEvent("mousedown"));

      expect(mousedown.value).toBe(0);

      b.dispatchEvent(createMouseEvent("mouseup"));

      expect(mouseup.value).toBe(1);
    });
  });

  test(`[onclick, onmousedown] => [onmouseup]`, () => {
    startRender((r) => {
      const click = eventCounter(Events.onClick);
      const mousedown = eventCounter(Events.onMouseDown);
      const mouseup = eventCounter(Events.onMouseUp);
      r(h.div().e([
        click.event,
        mousedown.event,
      ]));
      const b = r(h.div().e([
        mouseup.event,
      ]));

      b.dispatchEvent(createMouseEvent("click"));

      expect(click.value).toBe(0);

      b.dispatchEvent(createMouseEvent("mousedown"));

      expect(mousedown.value).toBe(0);

      b.dispatchEvent(createMouseEvent("mouseup"));

      expect(mouseup.value).toBe(1);
    });
  });

  test(`preventDefault method should trigger native prevent default behavior`, () => {
    startRender<HTMLInputElement>((r) => {
      const n = r(
        h.input("", { type: "checkbox" })
          .e(Events.onClick(() => EventFlags.PreventDefault)),
      );

      n.dispatchEvent(createMouseEvent("click"));
      expect(n.checked).toBeFalsy();
    });
  });

  test(`EventFlags.PreventDefault should trigger native prevent default behavior`, () => {
    startRender<HTMLInputElement>((r) => {
      const n = r(
        h.input("", { type: "checkbox" })
          .e(Events.onClick((e) => EventFlags.PreventDefault)),
      );

      n.dispatchEvent(createMouseEvent("click"));
      expect(n.checked).toBeFalsy();
    });
  });
});
