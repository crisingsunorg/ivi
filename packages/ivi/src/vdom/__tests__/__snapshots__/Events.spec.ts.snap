// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events update children 1`] = `2`;
